tests/components/sensor/test_init.py

```diff
@@ -2588,3 +2589,77 @@ def test_deprecated_constants_sensor_device_class(
     import_and_test_deprecated_constant_enum(
         caplog, sensor, enum, "DEVICE_CLASS_", "2025.1"
     )
+
+
+async def test_suggested_unit_guard(
+    hass: HomeAssistant,
+    caplog: pytest.LogCaptureFixture,
+) -> None:
+    """Test suggested_unit_of_measurement guard.
+
+    Only valid units should be stored in the entity registry.
+    On an invalid unit, a log entry should be created.
+    """
+    entity_registry = er.async_get(hass)
+    platform = getattr(hass.components, "test.sensor")
+    platform.init(empty=True)
+
+    state_value = 10
+    native_unit = UnitOfTemperature.CELSIUS
+    invalid_suggested_unit = "invalid_unit"
+    valid_suggested_unit = UnitOfTemperature.KELVIN
+
+    platform.ENTITIES["0"] = platform.MockSensor(
+        name="Invalid",
+        device_class=SensorDeviceClass.TEMPERATURE,
+        native_unit_of_measurement=native_unit,
+        suggested_unit_of_measurement=invalid_suggested_unit,
+        native_value=str(state_value),
+        unique_id="invalid",
+    )
+    entity_invalid = platform.ENTITIES["0"]
+
+    platform.ENTITIES["1"] = platform.MockSensor(
+        name="Valid",
+        device_class=SensorDeviceClass.TEMPERATURE,
+        native_unit_of_measurement=native_unit,
+        native_value=str(state_value),
+        suggested_unit_of_measurement=valid_suggested_unit,
+        unique_id="valid",
+    )
+    entity_valid = platform.ENTITIES["1"]
+
+    assert await async_setup_component(hass, "sensor", {"sensor": {"platform": "test"}})
+    await hass.async_block_till_done()
+
+    # Invalid unit should have native unit
+    state = hass.states.get(entity_invalid.entity_id)
+    assert int(state.state) == state_value
+    assert state.attributes[ATTR_UNIT_OF_MEASUREMENT] == native_unit
+
+    # Assert the suggested unit is NOT stored in the registry
+    entry = entity_registry.async_get(entity_invalid.entity_id)
+    assert entry.unit_of_measurement == invalid_suggested_unit
```

Comment:
In the entity registry still, the invalid unit will be set, but the state has the native one. Is this fine?